packages/ai/core/tool/mcp.ts

```diff
@@ -0,0 +1,209 @@
+}
+
+// Based off: https://github.com/dnakov/claude-code/blob/da716d77b251868918bb5776377f85a18a47ffdf/src/services/mcpClient.ts#L224
+async function connectToServer(
+  name: string,
+  serverConfig: McpServerConfig,
+): Promise<Client> {
+  const transport =
+    serverConfig.type === 'sse'
+      ? new SSEClientTransport(new URL(serverConfig.url), serverConfig.options)
+      : new StdioClientTransport({
+          ...serverConfig,
+          env: {
+            ...process.env,
+            ...serverConfig.env,
+          } as Record<string, string>,
+        });
+
+  const client = new Client(
+    {
+      name: `${name}-client`,
+      version: '0.1.0',
+    },
+    {
+      capabilities: {},
+    },
+  );
+
+  const CONNECTION_TIMEOUT_MS = 5000;
+  const connectPromise = client.connect(transport);
+  const timeoutPromise = new Promise<never>((_, reject) => {
+    const timeoutId = setTimeout(() => {
+      reject(
+        new Error(
+          `Connection to MCP server "${name}" timed out after ${CONNECTION_TIMEOUT_MS}ms`,
+        ),
+      );
+    }, CONNECTION_TIMEOUT_MS);
+
+    connectPromise.then(
+      () => clearTimeout(timeoutId),
+      error => {
+        clearTimeout(timeoutId);
+        reject(error);
+      },
+    );
+  });
+
+  try {
+    await Promise.race([connectPromise, timeoutPromise]);
+  } catch (error) {
+    if (error instanceof Error) {
+      throw new Error(
+        `Failed to connect to MCP server "${name}". ${error.message}`,
+        { cause: error },
+      );
+    }
+    throw error;
+  }
+
+  return client;
+}
+
+// To address:
+
+/**
+ * Refreshing tools:
+ * 
+ * - Would it be better to ship a global "tool/client manager" class? 
+ * - Gives users more granular control over clients, refreshing tools
+ * (e.g. getTools, refreshTools, closeConnections)
+ * - Even in non-serverless environments, we need a way to refresh tools (custom Transport?)
+ * 
+ * const toolManager = new McpToolManager({
+  sse: {
+    type: 'sse',
+    url: 'https://server.com/sse'
+  },
+  local: {
+    type: 'stdio',
+    command: 'npx'
+  }
+});
+ */
+
+/**
+ * Get Lars review on:
+ * - Generic type approach for dynamic input parameters
+ * - Which package to add to? Core is okie?
+ */
+
+/**
+ * Add tests, web socket impl. for `msw`
+ * Read: https://mswjs.io/docs/basics/handling-websocket-events
+ * Also add new example(s) for MCP stdio (and SSE?)
+ */
+
+/**
+ * Overall improvement of error handling
+ */
+
+export async function createMcpTools<
+  TToolSchemas extends McpToolSchemaMap = {},
+>(config: CreateMcpToolsConfig): Promise<McpToolset<TToolSchemas>> {
+  const toolSet: McpToolset<TToolSchemas> = {
+    tools: {} as McpToolset<TToolSchemas>['tools'],
+    clients: {},
+  };
+
+  const clientConnections = await Promise.allSettled(
+    Object.entries(config.servers).map(async ([name, serverConfig]) => {
+      const client = await connectToServer(name, serverConfig);
+      return { name, client };
+    }),
+  );
```

Comment:
I'm confused - does a single McpToolSet not match a single client/server connection?

My understanding was 1 McpToolSet <> 1 McpClient <> 1 McpServer - is that incorrect?

Reply:
The conversion tool accepts multiple server configs, spawns a client for each one, and combines all available tools; but I'm open to being opinionated about focusing on a single server<>client relationship - lmk what you think!

Reply:
Follow-up: converted to enforcing a single server; users can invoke the tool N times for every server they'd like to pull tools from!